connectivity/FEATURE_BLE/source/generic/GapImpl.cpp

```diff
@@ -2239,43 +2178,274 @@ void Gap::on_extended_advertising_report(
+            event.getPeerAddress(),
+            &peer_address_type,
+            &peer_address
+        );
+
+        if (address_resolved) {
+            conclude_signal_connection_complete_after_address_resolution(
+                event,
+                peer_address_type,
+                peer_address
+            );
+            return;
+        }
+    }
+
+    /* if successful then proceed to call the handler immediately same as for when privacy is disabled */
+    if (address_resolved) {
+        _event_handler->onConnectionComplete(
+            event
+        );
+    } else {
+        _address_registry.queue_resolve_address(event.getPeerAddress());
+
+        PendingConnectionCompleteEvent *pending_event = new(std::nothrow) PendingConnectionCompleteEvent(
+            event
+        );
+
+        bool allocation_successful = false;
+        if (pending_event && pending_event->event) {
+            allocation_successful = _connections_pending_address_resolution.push(pending_event);
+        }
+        if (!allocation_successful) {
+            event.setStatus(BLE_ERROR_NO_MEM);
+            _event_handler->onConnectionComplete(
+                event
+            );
+        }
+    }
+#else
+    _event_handler->onConnectionComplete(
+            event
+    );
 #endif // BLE_FEATURE_PRIVACY
+}
 
-    if (!_event_handler) {
-        return;
+void Gap::conclude_signal_connection_complete_after_address_resolution(
+    ConnectionCompleteEvent &event,
+    target_peer_address_type_t identity_address_type,
+    const address_t *identity_address
+)
+{
+#if BLE_ROLE_PERIPHERAL
+    bool resolvable_address_not_known = false;
+#endif // BLE_ROLE_PERIPHERAL
+    /* fix the event addresses */
+    if (identity_address) {
+        /* move old address to resolvable address */
+        event.setPeerResolvablePrivateAddress(event.getPeerAddress());
+        event.setLocalResolvablePrivateAddress(_address_registry.get_resolvable_private_address());
+
+        event.setPeerAddress(*identity_address);
+        event.setPeerAddressType((peer_address_type_t&)identity_address_type);
     }
+#if BLE_ROLE_PERIPHERAL
+    if (!identity_address) {
+        if (_peripheral_privacy_configuration.resolution_strategy ==
+            peripheral_privacy_configuration_t::REJECT_NON_RESOLVED_ADDRESS) {
+            // Reject connection request - the user will get notified through a callback
+            _pal_gap.disconnect(
+                event.getConnectionHandle(),
+                local_disconnection_reason_t::AUTHENTICATION_FAILURE
+            );
+            return;
+        }
+        resolvable_address_not_known = true;
+    }
+#endif // BLE_ROLE_PERIPHERAL
+
+    _event_handler->onConnectionComplete(
+        event
+    );
+#if BLE_ROLE_PERIPHERAL
+#if BLE_FEATURE_SECURITY
+    if (resolvable_address_not_known) {
+        ble::SecurityManager &sm = createBLEInstance()->getSecurityManager();
```

Comment:
Please use `BLE::Instance().securityManager()` to access the security manager. It is less misleading than `createBLEInstance`.